rust/basic_bitcoin/src/basic_bitcoin/src/bitcoin_wallet.rs

```diff
@@ -204,20 +249,47 @@ async fn sign_transaction(
         sig_with_hashtype.push(SIG_HASH_TYPE.as_u32() as u8);
         input.script_sig = Builder::new()
             .push_slice(sig_with_hashtype.as_slice())
-            .push_slice(public_key.as_slice())
+            .push_slice(own_public_key)
             .into_script();
         input.witness.clear();
     }
 
     transaction
 }
 
-fn sha256(data: Vec<u8>) -> Vec<u8> {
+fn sha256(data: &[u8]) -> Vec<u8> {
     let mut hasher = sha2::Sha256::new();
     hasher.update(data);
     hasher.finalize().to_vec()
 }
 
+// Converts a public key to a P2PKH address.
+fn public_key_to_p2pkh_address(network: Network, public_key: &[u8]) -> String {
+    // sha256 + ripmd160
+    let mut hasher = ripemd::Ripemd160::new();
+    hasher.update(sha256(public_key));
+    let result = hasher.finalize();
+
+    let prefix = match network {
+        Network::Testnet | Network::Regtest => 0x6f,
+        Network::Mainnet => 0x00,
+    };
+    let mut data_with_prefix = vec![prefix];
+    data_with_prefix.extend(result);
+
+    let checksum = &sha256(&sha256(&data_with_prefix.clone()))[..4];
+
+    let mut full_address = data_with_prefix;
+    full_address.extend(checksum);
+
+    bs58::encode(full_address).into_string()
+}
+
+// A mock for rubber-stamping ECDSA signatures.
+async fn mock_signer(_derivation_path: Vec<Vec<u8>>, _message_hash: Vec<u8>) -> Vec<u8> {
+    vec![0; 64]
```

Comment:
```suggestion
    vec![1; 64]
```

It seems that your answer:

> It's actually important. The signature from ECDSA is always 64 bytes. Converting to DER is taken care of in the sec1_to_der method, so the current logic takes that into account. No need for anything else here.

disappeared, I don't know if that was intentional.

Indeed the size varies when serializing in DER format.

What I am not happy with (but is easily solvable) is that we pass `vec![0; 64]` from `mock_signer` to `sec1_to_der` to know how many bytes the signed transaction will be. However in production we don't know if `sec1_signature[0] & 0x80 != 0` and `sec1_signature[32] & 0x80 != 0` will stand.
While for `vec![0; 64]` both conditions don't stand so we proceed to the `else` code blocks for both conditions and the problem is that these two blocks result in 2 less bytes (per input) than the `if` code blocks, so here in production the transaction size may be 2 more bytes long per input.

This detail looks important to me as a user providing the minimal fee per byte (1 Satoshi per byte IIRC) then because the signed transaction size is under-estimated then the transaction will be considered as invalid by the Bitcoin network.

The easy solution is to pass `vec![1; 64]` from `mock_signer` to `sec1_to_der` instead of `vec![0; 64]`, likewise both conditions will stand.

Reply:
Yes, exactly, that will make the DER signature to be the longest form. That implies that we'll sometimes be paying a slightly higher fee than we need to, but I think that's ok. The difference is very small.